File: domain/src/entities/serviceplan.rs
use diesel_derive_newtype::*;
use newtype_derive::*;
use serde_derive::{Deserialize, Serialize};
use uuid::Uuid;

use crate::*;
use schema::service_plans;

#[derive(
    Clone,
    Debug,
    PartialEq,
    Serialize,
    Deserialize,
    // Queryable,
    // Identifiable,
    // Insertable,
    // AsChangeset,
)]
#[repr(i64)]
pub enum ServiceAlert {
    NoService,
    Days(i32),
    Time(i32),
    Distance(i32),
    Climb(i32),
    Descend(i32),
    Count(i32),
}

#[derive(
    DieselNewType, Clone, Copy, Debug, Hash, PartialEq, Eq, Serialize, Deserialize, Default,
)]
pub struct ServicePlanId(Uuid);

NewtypeDisplay! { () pub struct ServicePlanId(); }
NewtypeFrom! { () pub struct ServicePlanId(Uuid); }

impl ServicePlanId {
    pub(crate) fn new() -> Self {
        Uuid::now_v7().into()
    }

    async fn get(self, store: &mut impl ServicePlanStore) -> TbResult<ServicePlan> {
        store.get(self).await
    }

    pub async fn delete(self, user: &dyn Person, store: &mut impl Store) -> TbResult<Vec<Service>> {
        let plan = self.get(store).await?;
        plan.part.checkuser(user, store).await?;

        let res = Service::reset_plan(self, store).await?;

        // delete service
        ServicePlanStore::delete(store, self).await?;
        Ok(res)
    }
    /* async fn check(
        self,
        last_service: OffsetDateTime,
        usage: &Usage,
        store: &mut impl ServicePlanStore,
    ) -> TbResult<ServiceAlert> {
        use time::{Duration, OffsetDateTime};
        use ServiceAlert::*;
        let plan = self.get(store).await?;
        if let Some(plan) = plan.days {
            if Duration::days((plan).into()) <= OffsetDateTime::now_utc() - last_service {
                return Ok(Time(plan));
            };
        }
        if let Some(plan) = plan.time {
            if plan <= usage.time {
                return Ok(Time(plan));
            };
        }
        if let Some(plan) = plan.climb {
            if plan <= usage.climb {
                return Ok(Climb(plan));
            };
        }
        if let Some(plan) = plan.descend {
            if plan <= usage.descend {
                return Ok(Descend(plan));
            };
        }
        if let Some(plan) = plan.distance {
            if plan <= usage.distance {
                return Ok(Distance(plan));
            };
        }
        if let Some(plan) = plan.count {
            if plan <= usage.count {
                return Ok(Count(plan));
            };
        }
        Ok(NoService)
    } */
}

///
///
#[derive(
    Clone,
    Debug,
    PartialEq,
    Serialize,
    Deserialize,
    Queryable,
    Identifiable,
    Insertable,
    AsChangeset,
)]
pub struct ServicePlan {
    #[serde(default = "ServicePlanId::new")]
    pub id: ServicePlanId,
    /// the gear or part involved
    /// if hook is None the plan is for a specific part
    /// if it's Some(hook) it is a generic plan for that hook
    part: PartId,
    /// This is only really used for generic plans
    /// for a specific part it is set to the PartType of the part
    what: PartTypeId,
    /// where it is attached
    hook: Option<PartTypeId>,
    name: String,
    /// Time until service
    pub days: Option<i32>,
    /// Usage time
    pub time: Option<i32>,
    /// Usage distance
    pub distance: Option<i32>,
    /// Overall climbing
    pub climb: Option<i32>,
    /// Overall descending
    pub descend: Option<i32>,
    /// number of activities
    pub count: Option<i32>,
}

impl ServicePlan {
    pub async fn create(
        mut self,
        user: &dyn Person,
        store: &mut (impl ServicePlanStore + PartStore),
    ) -> TbResult<Self> {
        self.part.checkuser(user, store).await?;
        self.id = ServicePlanId::new();
        store.create(self).await
    }

    pub async fn update(
        mut self,
        user: &dyn Person,
        store: &mut (impl ServicePlanStore + PartStore),
    ) -> TbResult<ServicePlan> {
        let plan = self.id.get(store).await?;
        plan.part.checkuser(user, store).await?;
        // You cannot change these
        self.part = plan.part;
        self.what = plan.what;
        self.hook = plan.hook;
        store.update(self).await
    }

    pub(crate) async fn for_part(
        part: PartId,
        store: &mut impl ServicePlanStore,
    ) -> TbResult<Vec<Self>> {
        store.by_part(part).await
    }
}

File: frontend/src/ServicePlan/serviceplan.ts
import { get_days, handleError, myfetch } from "../lib/store";
import { mapable, type Map, filterValues, by } from "../lib/mapable";
import { Service, services } from "../Service/service";
import { Part } from "../Part/part";
import {
  part_at_hook,
  type Attachment,
  attachment_for_part,
} from "../Attachment/attachment";
import type { Usage } from "../Usage/usage";

const is_set = (n: number | null) => n != null && n > 0;

export class Limits {
  /// Time until service
  days: number | null;
  /// Usage time
  time: number | null;
  /// Usage distance
  distance: number | null;
  /// Overall climbing
  climb: number | null;
  /// Overall descending
  descend: number | null;
  /// number of activities
  rides: number | null;

  constructor(data: any) {
    this.days = Number(data.days) || null;
    this.time = Number(data.time) || null;
    this.distance = Number(data.distance) || null;
    this.climb = Number(data.climb) || null;
    this.descend = Number(data.descend) || null;
    this.rides = Number(data.rides) || null;
  }

  static keys: (
    | "days"
    | "time"
    | "distance"
    | "climb"
    | "descend"
    | "rides"
  )[] = ["days", "time", "distance", "climb", "descend", "rides"];

  valid() {
    return (
      is_set(this.days) ||
      is_set(this.time) ||
      is_set(this.distance) ||
      is_set(this.climb) ||
      is_set(this.descend) ||
      is_set(this.rides)
    );
  }
}

export class ServicePlan extends Limits {
  id?: string;
  /// the gear or part involved
  /// if hook is None the plan is for a specific part
  /// if it's Some(hook) it is a generic plan for that hook
  part: number;
  /// This is only really used for generic plans
  /// for a specific part it is set to the PartType of the part
  what: number;
  /// where it is attached
  hook: number | null;
  name: string;
  constructor(data: any) {
    super(data);
    this.id = data.id || "00000000-0000-0000-0000-000000000000";
    this.part = data.part;
    this.what = data.what;
    this.hook = data.hook || null;

    this.name = data.name || "";
  }

  async create() {
    return await myfetch("/api/plan", "POST", this)
      .then((data) => plans.updateMap([data]))
      .catch(handleError);
  }

  async update() {
    return await myfetch("/api/plan", "PUT", this)
      .then((data) => plans.updateMap([data]))
      .catch(handleError);
  }

  async delete() {
    await myfetch("/api/plan/" + this.id, "DELETE")
      .then((data) => services.updateMap(data))
      .catch(handleError);
    plans.deleteItem(this.id);
  }

  valid() {
    return super.valid() && this.name.length > 0 && this.what != undefined;
  }

  services(part: Part, services: Map<Service>) {
    return filterValues(
      services,
      // @ts-ignore
      (s) => s.part_id == part.id && s.plans.includes(this.id),
    ).sort(by("time"));
  }

  getpart(parts: Map<Part>, attaches: Map<Attachment>) {
    return parts[part_at_hook(this.part, this.what, this.hook, attaches)];
  }

  due(part: Part, service: Service | undefined, usages: Map<Usage>) {
    let res = new Limits({});
    let time = service ? service.time : part.purchase;
    let usage = usages[part.usage];
    if (service) usage = usage.sub(usages[service.usage]);
    if (this.days) res.days = this.days - get_days(time);
    if (this.time) res.time = this.time - Math.floor(usage.time / 3600);
    if (this.distance)
      res.distance = this.distance - Math.floor(usage.distance / 1000);
    if (this.climb) res.climb = this.climb - usage.climb;
    if (this.descend) res.descend = this.descend - usage.descend;
    if (this.rides) res.rides = this.rides - usage.count;
    return res;
  }
}

export function plans_for_gear(
  part: number | undefined,
  plans: Map<ServicePlan>,
  atts: Map<Attachment>,
  time = new Date(),
) {
  let att = attachment_for_part(part, atts, time);
  return filterValues(
    plans,
    (s) =>
      s.part == part ||
      (s.hook == att?.hook && s.part == att?.gear && s.what == att?.what),
  );
}

export function plans_for_part(
  part: Part,
  time: Date,
  plans: Map<ServicePlan>,
  atts: Map<Attachment>,
) {
  if (part.isGear())
    return filterValues(plans, (p) => p.part == part.id && p.hook == null);
  else return plans_for_gear(part.id, plans, atts, time);
}
export const plans = mapable("id", (s) => new ServicePlan(s));